The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4867054903/jobs/8679247024) [(plain)](https://github.com/rust-lang/rust/commit/3afe8cdfa8731abbcdde929ccee46ec367a4bc75/checks/13188829228/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 0.371 seconds
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 155 tests
iii..............iiiiiiiiiiiiiF.........................................................  88/155

failures:

---- [assembly] tests/assembly/align_offset.rs stdout ----
---- [assembly] tests/assembly/align_offset.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-14/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/align_offset/align_offset.s" "/checkout/tests/assembly/align_offset.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC" "--dump-input-context" "100"
stdout: none
/checkout/tests/assembly/align_offset.rs:31:11: error: CHECK: expected string not found in input
/checkout/tests/assembly/align_offset.rs:31:11: error: CHECK: expected string not found in input
// CHECK: orq
          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/align_offset/align_offset.s:144:6: note: scanning from here
 shrq $2, %rax
     ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/align_offset/align_offset.s:147:22: note: possible intended match here
 .size align_offset_word_ptr, .Lfunc_end4-align_offset_word_ptr

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/align_offset/align_offset.s
Check file: /checkout/tests/assembly/align_offset.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
            .
            .
            .
            .
           44:  subq %rdi, %rcx 
           45:  movq %r9, %rax 
           46:  shrq %rax 
           47:  leaq .L__unnamed_1(%rip), %rdx 
           48:  movzbl (%rax,%rdx), %eax 
           49:  cmpq $17, %rsi 
           50:  jb .LBB0_9 
           51:  movq %r9, %rdi 
           52:  imulq %rax, %rdi 
           53:  movl $2, %r10d 
           54:  movl $2, %edx 
           55:  subq %rdi, %rdx 
           56:  imulq %rdx, %rax 
           57:  cmpq $257, %rsi 
           58:  jb .LBB0_9 
           59:  movq %rax, %rdx 
           60:  imulq %r9, %rdx 
           61:  subq %rdx, %r10 
           62:  imulq %r10, %rax 
           63:  cmpq $65537, %rsi 
           64:  jb .LBB0_9 
           65:  movq %rax, %rdi 
           66:  imulq %r9, %rdi 
           67:  movl $2, %r10d 
           68:  movl $2, %edx 
           69:  subq %rdi, %rdx 
           70:  imulq %rdx, %rax 
           71:  movabsq $4294967297, %rdx 
           72:  cmpq %rdx, %rsi 
           73:  jb .LBB0_9 
           74:  imulq %rax, %r9 
           75:  subq %r9, %r10 
           76:  imulq %r10, %rax 
           77: .LBB0_9: 
           78:  andq %r8, %rax 
           79:  imulq %rcx, %rax 
           80:  andq %r8, %rax 
           81:  retq 
           82: .Lfunc_end0: 
           83:  .size _ZN4core3ptr12align_offset17h663c57cfe37f2997E, .Lfunc_end0-_ZN4core3ptr12align_offset17h663c57cfe37f2997E 
           84:  .cfi_endproc 
           85:  
           86:  .section .text._ZN4core3ptr12align_offset17hdc1f1e619dd1edd9E,"ax",@progbits 
           87:  .globl _ZN4core3ptr12align_offset17hdc1f1e619dd1edd9E 
           88:  .p2align 4, 0x90 
           89:  .type _ZN4core3ptr12align_offset17hdc1f1e619dd1edd9E,@function 
           90: _ZN4core3ptr12align_offset17hdc1f1e619dd1edd9E: 
           91:  .cfi_startproc 
           92:  leaq (%rsi,%rdi), %rax 
           93:  addq $-1, %rax 
           94:  negq %rsi 
           95:  andq %rsi, %rax 
           96:  subq %rdi, %rax 
           97:  retq 
           98: .Lfunc_end1: 
           99:  .size _ZN4core3ptr12align_offset17hdc1f1e619dd1edd9E, .Lfunc_end1-_ZN4core3ptr12align_offset17hdc1f1e619dd1edd9E 
          100:  .cfi_endproc 
          101:  
          102:  .section .text.align_offset_byte_ptr,"ax",@progbits 
          103:  .globl align_offset_byte_ptr 
          104:  .p2align 4, 0x90 
          105:  .type align_offset_byte_ptr,@function 
          106: align_offset_byte_ptr: 
          107:  .cfi_startproc 
          108:  leaq 31(%rdi), %rax 
          109:  andq $-32, %rax 
          110:  subq %rdi, %rax 
          111:  retq 
          112: .Lfunc_end2: 
          113:  .size align_offset_byte_ptr, .Lfunc_end2-align_offset_byte_ptr 
          114:  .cfi_endproc 
          115:  
          116:  .section .text.align_offset_byte_slice,"ax",@progbits 
          117:  .globl align_offset_byte_slice 
          118:  .p2align 4, 0x90 
          119:  .type align_offset_byte_slice,@function 
          121:  .cfi_startproc 
          121:  .cfi_startproc 
          122:  leaq 31(%rdi), %rax 
          123:  andq $-32, %rax 
          124:  subq %rdi, %rax 
          125:  retq 
          126: .Lfunc_end3: 
          127:  .size align_offset_byte_slice, .Lfunc_end3-align_offset_byte_slice 
          128:  .cfi_endproc 
          129:  
          130:  .section .text.align_offset_word_ptr,"ax",@progbits 
          131:  .globl align_offset_word_ptr 
          132:  .p2align 4, 0x90 
          133:  .type align_offset_word_ptr,@function 
          134: align_offset_word_ptr: 
          135:  .cfi_startproc 
          136:  movq $-1, %rax 
          137:  testb $3, %dil 
          138:  je .LBB4_1 
          139:  retq 
          140: .LBB4_1: 
          141:  leaq 31(%rdi), %rax 
          142:  andq $-32, %rax 
          143:  subq %rdi, %rax 
          144:  shrq $2, %rax 
check:31'0          X~~~~~~~~~ error: no match found
          145:  retq 
check:31'0     ~~~~~~
          146: .Lfunc_end4: 
check:31'0     ~~~~~~~~~~~~~
          147:  .size align_offset_word_ptr, .Lfunc_end4-align_offset_word_ptr 
check:31'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:31'1                          ?                                           possible intended match
          148:  .cfi_endproc 
check:31'0     ~~~~~~~~~~~~~~
          149:  
check:31'0     ~
          150:  .section .text.align_offset_word_slice,"ax",@progbits 
check:31'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          151:  .globl align_offset_word_slice 
          152:  .p2align 4, 0x90 
          153:  .type align_offset_word_slice,@function 
          154: align_offset_word_slice: 
          155:  .cfi_startproc 
          156:  leaq 31(%rdi), %rax 
          157:  andq $-32, %rax 
          158:  subq %rdi, %rax 
          159:  shrq $2, %rax 
          160:  retq 
          161: .Lfunc_end5: 
          162:  .size align_offset_word_slice, .Lfunc_end5-align_offset_word_slice 
          163:  .cfi_endproc 
          164:  
          165:  .type .L__unnamed_1,@object 
          166:  .section .rodata.cst8,"aM",@progbits,8 
          167: .L__unnamed_1: 
          168:  .ascii "\001\013\r\007\t\003\005\017" 
          169:  .size .L__unnamed_1, 8 
          170:  
          171:  .section ".note.GNU-stack","",@progbits 
------------------------------------------



```

</details>

--- a/library/core/src/lib.rs
+++ b/library/core/src/lib.rs
@@ -107,6 +107,7 @@
 #![feature(const_arguments_as_str)]
 #![feature(const_array_from_ref)]
 #![feature(const_array_into_iter_constructors)]
+#![feature(const_assume)]
 #![feature(const_bigint_helper_methods)]
 #![feature(const_black_box)]
 #![feature(const_caller_location)]

--- a/library/core/src/ptr/mod.rs
+++ b/library/core/src/ptr/mod.rs
@@ -1632,8 +1632,8 @@ pub(crate) const unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usiz
     // FIXME(#75598): Direct use of these intrinsics improves codegen significantly at opt-level <=
     // 1, where the method versions of these operations are not inlined.
     use intrinsics::{
-        cttz_nonzero, exact_div, mul_with_overflow, unchecked_rem, unchecked_shl, unchecked_shr,
-        unchecked_sub, wrapping_add, wrapping_mul, wrapping_sub,
+        assume, cttz_nonzero, exact_div, mul_with_overflow, unchecked_rem, unchecked_shl,
+        unchecked_shr, unchecked_sub, wrapping_add, wrapping_mul, wrapping_sub,
     };
 
     /// Calculate multiplicative modular inverse of `x` modulo `m`.
@@ -1724,12 +1724,18 @@ pub(crate) const unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usiz
         // in a branch-free way and then bitwise-OR it with whatever result the `-p mod a`
         // computation produces.
 
+        let aligned_address = wrapping_add(addr, a_minus_one) & wrapping_sub(0, a);
+        let byte_offset = wrapping_sub(aligned_address, addr);
+        // FIXME: Remove the assume after <https://github.com/llvm/llvm-project/issues/62502>
+        // SAFETY: Masking by `-a` can only affect the low bits, and thus cannot have reduced
+        // the value by more than `a-1`, so even though the intermediate values might have
+        // wrapped, the byte_offset is always in `[0, a)`.
+        unsafe { assume(byte_offset < a) };
+
         // SAFETY: `stride == 0` case has been handled by the special case above.
         let addr_mod_stride = unsafe { unchecked_rem(addr, stride) };
 
         return if addr_mod_stride == 0 {
-            let aligned_address = wrapping_add(addr, a_minus_one) & wrapping_sub(0, a);
-            let byte_offset = wrapping_sub(aligned_address, addr);
             // SAFETY: `stride` is non-zero. This is guaranteed to divide exactly as well, because
             // addr has been verified to be aligned to the original type’s alignment requirements.
             unsafe { exact_div(byte_offset, stride) }

--- a/tests/codegen/align-offset.rs
+++ b/tests/codegen/align-offset.rs
@@ -1,0 +1,78 @@
+// compile-flags: -O
+// min-llvm-version: 15.0 (because we're using opaque pointers)
+// ignore-debug (debug assertions in `slice::from_raw_parts` block optimizations)
+
+#![crate_type = "lib"]
+
+// CHECK-LABEL: @align8
+#[no_mangle]
+pub fn align8(p: *const u8) -> bool {
+    // CHECK: ret i1 true
+    p.align_offset(8) < 8
+}
+
+#[repr(align(4))]
+pub struct Align4([u8; 4]);
+
+// CHECK-LABEL: @align_to4
+#[no_mangle]
+pub fn align_to4(x: &[u8]) -> bool {
+    // CHECK: ret i1 true
+    let (prefix, _middle, suffix) = unsafe { x.align_to::<Align4>() };
+    prefix.len() < 4 && suffix.len() < 4
+}
+
+// CHECK-LABEL: @align_offset_byte_ptr(ptr{{.+}}%ptr)
+#[no_mangle]
+pub fn align_offset_byte_ptr(ptr: *const u8) -> usize {
+    // CHECK: %[[ADDR:.+]] = ptrtoint ptr %ptr to [[USIZE:i[0-9]+]]
+    // CHECK: %[[UP:.+]] = add [[USIZE]] %[[ADDR]], 31
+    // CHECK: %[[ALIGNED:.+]] = and [[USIZE]] %[[UP]], -32
+    // CHECK: %[[OFFSET:.+]] = sub [[USIZE]] %[[ALIGNED]], %[[ADDR]]
+
+    // Since we're offsetting a byte pointer, there's no further fixups
+    // CHECK-NOT: shr
+    // CHECK-NOT: div
+    // CHECK-NOT: select
+
+    // CHECK: ret [[USIZE]] %[[OFFSET]]
+    ptr.align_offset(32)
+}
+
+// CHECK-LABEL: @align_offset_word_slice(ptr{{.+}}align 4{{.+}}%slice.0
+#[no_mangle]
+pub fn align_offset_word_slice(slice: &[Align4]) -> usize {
+    // CHECK: %[[ADDR:.+]] = ptrtoint ptr %slice.0 to [[USIZE]]
+    // CHECK: %[[UP:.+]] = add [[USIZE]] %[[ADDR]], 31
+    // CHECK: %[[ALIGNED:.+]] = and [[USIZE]] %[[UP]], -32
+    // CHECK: %[[BOFFSET:.+]] = sub [[USIZE]] %[[ALIGNED]], %[[ADDR]]
+    // CHECK: %[[OFFSET:.+]] = lshr exact [[USIZE]] %[[BOFFSET]], 2
+
+    // Slices are known to be aligned, so we don't need the "maybe -1" path
+    // CHECK-NOT: select
+
+    // CHECK: ret [[USIZE]] %[[OFFSET]]
+    slice.as_ptr().align_offset(32)
+}
+
+
+// CHECK-LABEL: @align_offset_word_ptr(ptr{{.+}}%ptr
+#[no_mangle]
+pub fn align_offset_word_ptr(ptr: *const Align4) -> usize {
+    // CHECK: %[[ADDR:.+]] = ptrtoint ptr %ptr to [[USIZE]]
+    // CHECK: %[[UP:.+]] = add [[USIZE]] %[[ADDR]], 31
+    // CHECK: %[[ALIGNED:.+]] = and [[USIZE]] %[[UP]], -32
+    // CHECK: %[[BOFFSET:.+]] = sub [[USIZE]] %[[ALIGNED]], %[[ADDR]]
+
+    // While we can always get a *byte* offset that will work, if the original
+    // pointer is unaligned it might be impossible to return an *element* offset
+    // that will make it aligned. We want it to be a `select`, not a `br`, so
+    // that the assembly will be branchless.
+    // CHECK: %[[LOW:.+]] = and [[USIZE]] %[[ADDR]], 3
+    // CHECK: %[[ORIGINAL_ALIGNED:.+]] = icmp eq [[USIZE]] %[[LOW]], 0
+    // CHECK: %[[OFFSET:.+]] = lshr exact [[USIZE]] %[[BOFFSET]], 2
+    // CHECK: %[[R:.+]] = select i1 %[[ORIGINAL_ALIGNED]], [[USIZE]] %[[OFFSET]], [[USIZE]] -1
+
+    // CHECK: ret [[USIZE]] %[[R]]
+    ptr.align_offset(32)
+}

